src/main/java/org/opensearch/neuralsearch/processor/ScoreCombiner.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.neuralsearch.processor;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Objects;
+import java.util.PriorityQueue;
+import java.util.stream.Collectors;
+
+import lombok.extern.log4j.Log4j2;
+
+import org.apache.lucene.search.ScoreDoc;
+import org.apache.lucene.search.TopDocs;
+import org.apache.lucene.search.TotalHits;
+import org.opensearch.neuralsearch.search.CompoundTopDocs;
+
+/**
+ * Abstracts combination of scores in query search results.
+ */
+@Log4j2
+public class ScoreCombiner {
+
+    private static final Float ZERO_SCORE = 0.0f;
+
+    /**
+     * Performs score combination based on input combination technique. Mutates input object by updating combined scores
+     * @param queryTopDocs query results that need to be normalized, mutated by method execution
+     * @param combinationTechnique exact combination method that should be applied
+     * @return list of max combined scores for each shard
+     */
+    public List<Float> combineScores(final CompoundTopDocs[] queryTopDocs, final ScoreCombinationTechnique combinationTechnique) {
```

Comment:
Couple of comments on this function:
1. This function is long and has a lot of moving pieces. Lets break it into smaller pieces
2. Let's try to add inline comments on what we are trying to do. Like when scores are getting updated, when the scores are getting combined etc. This will ensure better readability.
3. I would like to add comments in a step by step format like we say we need to do 4 things and then mention inline when those steps are performed,

Reply:
ack